### PR TITLE
feat(execute/table): Refactored the table builder interfaces to support null value creation.

### PR DESCRIPTION
**Pull request by aanthony1243**
### Done checklist
- [X] docs/SPEC.md updated
- [X] Test cases written

### CODE REVIEW

File: functions/transformations/difference.go [hunk 1]
```diff
@@ -187,7 +187,7 @@ func (t *differenceTransformation) Process(id execute.DatasetID, tbl flux.Table)
 				d := differences[j]
 				switch c.Type {
 				case flux.TBool:
-					if err := arrow.AppendBools(builder, j, arrow.BoolSlice(cr.Bools(j), firstIdx, l)); err != nil {
+					if err := builder.AppendBools(j, arrow.BoolSlice(cr.Bools(j), firstIdx, l)); err != nil {
```
Comment (jsternberg) [inline functions/transformations/difference.go]: I'm not sure how this got in originally, but it's a memory leak. You have to release the slice after creating it.

File: functions/transformations/keys.go [hunk 2]
```diff
@@ -167,8 +168,7 @@ func (t *keysTransformation) Process(id execute.DatasetID, tbl flux.Table) error
 		}
 	}
 
-	// Append the keys to the column index.
-	if err := builder.AppendStrings(colIdx, keys); err != nil {
+	if err := builder.AppendStrings(colIdx, arrow.NewString(keys, nil)); err != nil {
```
Comment (jsternberg) [inline functions/transformations/keys.go]: This is a memory leak.

File: execute/table.go [hunk 3]
```diff
@@ -244,54 +228,36 @@ func AppendRecord(i int, cr flux.ColReader, builder TableBuilder) error {
...
-		case flux.TTime:
-			var val Time
-			if colMap[j] >= 0 {
-				val = values.Time(cr.Times(colMap[j]).Value(i))
+		var val values.Value
+		if colMap[j] >= 0 {
+			val = ValueForRow(cr, i, colMap[j])
+			err = builder.AppendValue(j, val)
+		} else {
+			switch c.Type {
```
Comment (jsternberg) [inline execute/table.go]: Is there a reason why `AppendNil` can't be used in this section?

File: functions/transformations/data_test.go [hunk 4]
```diff
@@ -95,12 +96,12 @@ func init() {
 		}
 	}
 
-	_ = normalTableBuilder.AppendTimes(0, times)
-	_ = normalTableBuilder.AppendTimes(1, startTimes)
-	_ = normalTableBuilder.AppendTimes(2, stopTimes)
-	_ = normalTableBuilder.AppendFloats(3, normalValues)
-	_ = normalTableBuilder.AppendStrings(4, t1)
-	_ = normalTableBuilder.AppendStrings(5, t2)
+	_ = normalTableBuilder.AppendTimes(0, arrow.NewInt(times, nil))
```
Comment (jsternberg) [inline functions/transformations/data_test.go]: These are also memory leaks.

**Comment by affo** [comment]
@aanthony1243 @jsternberg 

Hey guys, I am not against this change as discussed in Slack. I leave the discussion to you and follow the thread.

However, @jsternberg I can't see the point of having utilities to convert a call that uses go slices to one that uses arrow arrays, given that we should evolve to an arrow-oriented implementation and avoid the use of slices. Am I missing something?   
Are you suggesting that we perform this step post-alpha and, for now, we only fix utils?

At the moment, I wrote down a fix to utils in #777   
However, from the `arrow` package in flux, I can't write, like @aanthony1243 did:

```
col := b.cols[j].(*intColumnBuilder)
nullOffset := len(col.data)
```

And use the offset of that specific column to set nils. So, I must append values one by one.  
Should we move those utils into `execute`?

**Comment by aanthony1243** [comment]
I see the issues and I have some ideas.  we can talk offline and then update this PR.

**Comment by affo** [comment]
For difference memory leaks, see #792

**Comment by jsternberg** [comment]
I talked with @aanthony1243 offline and both of us seemed to be in agreement that the current table builder API needed to be revisited and we would revisit it after the alpha.

There were a  few points that are important:

1. The arrow builder should be exposed directly.
2. The utility methods are a bit of an anti-pattern since nobody knows they exist.
3. The table builder being an interface is likely an anti-pattern.

I'll give this another once over but let's just merge it so we have null support for the alpha and we'll revisit it later. We're not freezing any Go APIs yet anyway.

File: execute/table.go (hunk 3 above)
Comment (aanthony1243) [reply execute/table.go]: nope. I will fix it.

File: functions/transformations/difference.go (hunk 1 above)
Comment (aanthony1243) [reply functions/transformations/difference.go]: will let this get fixed in a currently open PR for difference.

**Pull request merged**